l10n_co_res_partner/models/l10n_co_res_partner.py

```diff
@@ -0,0 +1,459 @@
+            self.name2 = ''
+            self.lastname1 = ''
+            self.lastname2 = ''
+            self.pn_retri = 7
+        elif self.personType is 1:
+            self.companyName = False
+            self.pn_retri = False
+
+    @api.onchange('doctype')
+    def onChangeDocumentType(self):
+        """
+        If Document Type changes we delete the document number as for different
+        document types there are different rules that apply e.g. foreign
+        documents (e.g. 21) allows letters in the value. Here we reduce the
+        risk of having corrupt information about the contact.
+        @return: void
+        """
+        self.xidentification = False
+
+    @api.onchange('company_type')
+    def onChangeCompanyType(self):
+        """
+        This function changes the person type once the company type changes.
+        If it is a company, document type 31 will be selected automatically as
+        in Colombia it's more likely that it will be chosen by the user.
+        @return: void
+        """
+        if self.company_type == 'company':
+            self.personType = 2
+            self.is_company = True
+            self.doctype = 31
+        else:
+            self.personType = 1
+            self.is_company = False
+            self.doctype = 1
+
+    @api.onchange('is_company')
+    def onChangeIsCompany(self):
+        """
+        This function changes the person type field and the company type if
+        checked / unchecked
+        @return: void
+        """
+        if self.is_company is True:
+            self.personType = 2
+            self.company_type = 'company'
+            self.xbirthday = False
+        else:
+            self.is_company = False
+            self.company_type = 'person'
+
+    @api.onchange('change_country')
+    def onChangeAddress(self):
+        """
+        This function changes the person type field and the company type if
+        checked / unchecked
+        @return: void
+        """
+        if self.change_country is True:
+            self.country_id = False
+            self.state_id = False
+            self.xcity = False
+
+    def _check_dv(self, nit):
+        """
+        Function to calculate the check digit (DV) of the NIT. So there is no
+        need to type it manually.
+        @param nit: Enter the NIT number without check digit
+        @return: String
+        """
+        if self.doctype != 31:
+            return str(nit)
+
+        nitString = '0'*(15-len(nit)) + nit
+        vl = list(nitString)
+        result = (
+            int(vl[0])*71 + int(vl[1])*67 + int(vl[2])*59 + int(vl[3])*53 +
+            int(vl[4])*47 + int(vl[5])*43 + int(vl[6])*41 + int(vl[7])*37 +
+            int(vl[8])*29 + int(vl[9])*23 + int(vl[10])*19 + int(vl[11])*17 +
+            int(vl[12])*13 + int(vl[13])*7 + int(vl[14])*3
+        ) % 11
+
+        if result in (0, 1):
+            return str(result)
+        else:
+            return str(11-result)
```

Comment:
The proper place to put this algorithm is include this logic as a PR to odoo in the base_vat module